### ISSUE TITLE
A util to intercept messages and do evals statically 

### ISSUE THREAD

**Issue by ahuang11** [anaconda/llm-eval#62]
In Lumen, I added a util to intercept OpenAI API inputs and store them into a database for later use (specifically for llm-eval) https://github.com/holoviz/lumen/pull/642/files#diff-c4cdb059c13610094aca0d0d6ec22db1ce596833fa756f2115b059e3b75eed6fR304-R306 because it was very hard and tedious to re-create the inputs by hand, e.g. https://github.com/holoviz-dev/llm-eval-lumen/pull/1/files#diff-0cc92d7f95ea5c631aaecc55fbdaaba638930197ade596f6aa3c03212a9f1eeeR1-R116

I'm wondering if it would be a good idea to migrate this functionality into llm-eval so that users can use a chat interface, like Panel, to [bootstrap eval yaml files](https://github.com/holoviz-dev/llm-eval-lumen/pull/1), rather than manually building these evals?

**Comment by sadasant** [comment]
I agree! Let's do this!

**Comment by ahuang11** [comment]
I don't think I wrote this issue very clearly, so let me try again.

As far as I understand, at the moment, one would need to hard code evals like this

```python
metadata:
  name: Fibonacci Sequence
prompt_sequence:
  - prompt: Create a Python function called `fib` that takes an integer `n` and returns the `n`th number in the Fibonacci sequence. Use type hints and docstrings.
    checks:
      - check_type: REGEX
        pattern: "def fib\\([a-zA-Z_]+\\: int\\) -> int\\:"
      - check_type: PYTHON_CODE_BLOCKS_PRESENT
  - prompt: Create a set of assertion statements that test the function.
    checks:
      - check_type: CONTAINS
        value: assert fib(
      - check_type: PYTHON_CODE_BLOCKS_PRESENT
      - check_type: PYTHON_CODE_BLOCK_TESTS
        code_setup: |
          import re
        code_tests:
          - |
            def verify_mask_emails_with_no_email_returns_original_string(code_blocks: list[str]) -> bool:
                value = 'This is a string with no email addresses'
                return mask_emails(value) == value
```

This could work straightforwardly for cases where there isn't much context / `previous_messages`, but it gets difficult to manage if the context is much longer, and/or if the inputs get [transformed by another lib](https://github.com/holoviz-dev/llm-eval-lumen/pull/1/files#diff-7d70024096ddc45fb42723d12ee20b149d8568231b7dfb1b959bb48ece53a241R3-R18), like instructor (which wraps response_model).

So, my hope with this issue is that what if the evals can be generated and filled out dynamically? For example

```python
metadata:
  name: Fibonacci Sequence
prompt_sequence:
  - prompt: Create a Python function called `fib` that takes an integer `n` and returns the `n`th number in the Fibonacci sequence. Use type hints and docstrings.
    checks:
        <MANUALLY FILL OUT, OR DEFAULT TO THE ASSISTANT'S RESPONSE AND THEN TWEAK>
```

To take it a step further, since the chat history, including the response, is intercepted and captured into the database, would it be possible to run checks statically (i.e. not call the OpenAI/LLM again), and just compare the checks vs response? That way, we can naturally chat with the LLM using a Panel Chat Interface / Anaconda Assistant, and then run diagnostics on the logs?
